### PR TITLE
Params for other ops

### PR DESCRIPTION
**Pull request by notoraptor**
Continues #5866 (cannot reopen that PR). Rebased and re-read. Adding op params for:
- GpuCumOp
- GPUA_mrg_uniform
- AdvancedIncSubtensor1 (+ set `check_input=False` for AdvancedSubtensor1)
- GpuAdvancedIncSubtensor1
- GpuAdvancedIncSubtensor1_dev20
- Images2Neibs
- GpuImages2Neibs
- compile.ops.Shape_i

I think it will not speed up TRAVIS so much, as changes are mainly in gpuarray module, but now we can end with this big old PR!

@nouiz @lamblin @abergeron

### CODE REVIEW

File: theano/compile/ops.py [hunk 1]
```diff
@@ -346,6 +346,18 @@ def __init__(self, i):
...
         self.i = i
 
+    # NB:
+    # 1) params_type is defined as a property to avoid
+    #    loop in Python import caused by importing theano.scalar below
+    #    when params_type is defined directly in class code.
+    # 2) We wrap scalar into ParamsType (instead of directly using scalar as op param)
+    #    to avoid Theano converting scalar param to constant that would be later
+    #    hardcoded as litteral in C code, making us loose all the advantages of
+    #    using params.
```
Comment (nouiz) [inline theano/compile/ops.py]: Why it get converted to constant? Maybe that should be fixed. Not mandatory for this PR.

File: theano/tensor/subtensor.py [hunk 2]
```diff
@@ -1872,10 +1873,13 @@ class AdvancedIncSubtensor1(Op):
...
     __props__ = ('inplace', 'set_instead_of_inc')
+    check_input = False
+    params_type = ParamsType(inplace=scal.bool,
+                             set_instead_of_inc=scal.int8)
 
     def __init__(self, inplace=False, set_instead_of_inc=False):
-        self.inplace = inplace
-        self.set_instead_of_inc = set_instead_of_inc
+        self.inplace = bool(inplace)
+        self.set_instead_of_inc = int(bool(set_instead_of_inc))
```
Comment (notoraptor) [inline theano/tensor/subtensor.py]: Fixed.

File: theano/gpuarray/rng_mrg.py [hunk 3]
```diff
@@ -168,40 +173,34 @@ def gpu_kernels(self, node, name):
...
-        size_t odims[%(ndim)s+1];
         size_t n_elements = 1;
         unsigned int n_streams;
         int must_alloc_sample = ((NULL == %(o_sample)s)
                 || !pygpu_GpuArray_Check((PyObject*)%(o_sample)s)
                 || !(%(o_sample)s->ga.flags & GA_C_CONTIGUOUS)
-                || (PyGpuArray_NDIM(%(o_sample)s) != %(ndim)s));
+                || (PyGpuArray_NDIM(%(o_sample)s) != %(params)s->ndim));
+
+        size_t* odims = (size_t*)malloc(%(params)s->ndim * sizeof(size_t));
```
Comment (notoraptor) [inline theano/gpuarray/rng_mrg.py]: Ok, I will test on Windows. From memory, the use of alloca() seems discouraged as there may be some compatibility issue (function is not POSIX, and is machine and compiler-dependent): http://man7.org/linux/man-pages/man3/alloca.3.html#NOTES .

File: theano/gpuarray/extra_ops.py [hunk 4]
```diff
@@ -21,9 +24,11 @@ class GpuCumOp(GpuKernelBase, Op):
     """
     SUPPORTED_NDIMS = 3
     __props__ = ('axis', 'mode')
+    params_type = ParamsType(axis=scalar.int32,
+                             context=gpu_context_type)
 
     def __init__(self, axis, mode='add'):
-        self.axis = axis if axis else 0
+        self.axis = int(axis) if axis is not None else 0
```
Comment (notoraptor) [inline theano/gpuarray/extra_ops.py]: Done.

File: theano/gpuarray/rng_mrg.py (hunk 3 above)
Comment (nouiz) [inline theano/gpuarray/rng_mrg.py]: So I'm not sure what to do. Maybe we should keep that for another PR to don't delay this PR. We should discuss that with @lamblin and @abergeron

File: theano/compile/ops.py (hunk 1 above)
Comment (notoraptor) [inline theano/compile/ops.py]: Ok. For info, the related code seems to be here: https://github.com/Theano/Theano/blob/master/theano/gof/cc.py#L685 .

File: theano/gpuarray/rng_mrg.py (hunk 3 above)
Comment (nouiz) [inline theano/gpuarray/rng_mrg.py]: I made an issue about that: https://github.com/Theano/Theano/issues/6094
